File: backend/app/api/flights.py
from datetime import date, datetime, timezone
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session
from ..db import get_db
from ..models import Airport, FareObservation
from ..config import get_settings
from ..schemas import FlightSearchParams
from ..services.providers.amadeus import AmadeusProvider
from ..services.normalizer import amadeus_to_flight
from ..services.fallback import demo_flights, demo_history
from ..services.analytics import price_rating

router = APIRouter(prefix="/api/v1", tags=["flights"])

@router.get("/airports")
def search_airports(q: str = "", db: Session = Depends(get_db)):
    q = q.strip().lower()
    stmt = select(Airport).order_by(Airport.city)
    if q:
        stmt = stmt.where(
            (Airport.code.ilike(f"%{q}%")) |
            (Airport.city.ilike(f"%{q}%")) |
            (Airport.name.ilike(f"%{q}%"))
        )
    rows = db.execute(stmt.limit(20)).scalars().all()
    return [{"code":a.code,"city":a.city,"name":a.name} for a in rows]

@router.get("/flights")
async def search_flights(
    from_: str = Query(alias="from", min_length=3, max_length=3),
    to: str = Query(min_length=3, max_length=3),
    date: date = Query(...),
    adults: int = Query(1, ge=1, le=9),
    max_results: int = Query(20, ge=1, le=50),
    db: Session = Depends(get_db),
):
    origin, destination = from_.upper(), to.upper()
    settings = get_settings()
    provider = AmadeusProvider(settings.amadeus_client_id, settings.amadeus_client_secret, settings.amadeus_base_url)
    flights = []
    provider_used = False
    if provider.enabled:
        try:
            payload = await provider.search(origin, destination, str(date), adults, max_results)
            flights = [amadeus_to_flight(x, origin, destination) for x in payload.get("data", [])]
            provider_used = bool(flights)
            if flights:
                for f in flights:
                    fare = f["fare"]
                    db.add(FareObservation(
                        observed_at=datetime.now(timezone.utc),
                        departure_date=date,
                        route=f"{origin}-{destination}",
                        airline_code=f["airlineCode"],
                        flight_number=f["flightNumber"],
                        source="Amadeus",
                        source_offer_id=f["id"],
                        currency="INR",
                        base_fare=fare["base"],
                        taxes=fare["taxes"],
                        fees=fare["fees"],
                        total_fare=fare["total"],
                        cabin=f["cabin"],
                        raw_payload=f.get("_raw"),
                        valid=True,
                    ))
                db.commit()
        except Exception:
            flights = []
    if not flights:
        flights = demo_flights(origin, destination, str(date), max_results)
    history = db.execute(
        select(FareObservation.total_fare)
        .where(FareObservation.route == f"{origin}-{destination}")
        .order_by(FareObservation.observed_at.desc()).limit(200)
    ).scalars().all()
    history = [float(x) for x in history]
    for f in flights:
        f["priceRating"] = price_rating(f["fare"]["total"], history)
        f.pop("_raw", None)
    return {"params":{"from":origin,"to":destination,"date":str(date)},"flights":flights,"count":len(flights),"source":"Amadeus" if provider_used else "Demo fallback"}

@router.get("/flights/{flight_id}")
def flight_detail(flight_id: str, db: Session = Depends(get_db)):
    row = db.execute(select(FareObservation).where(FareObservation.source_offer_id == flight_id).order_by(FareObservation.observed_at.desc())).scalars().first()
    if row:
        return {
            "id": flight_id, "airlineCode": row.airline_code, "flightNumber": row.flight_number,
            "from": row.route[:3], "to": row.route[4:], "departTime": datetime.combine(row.departure_date, datetime.min.time(), tzinfo=timezone.utc),
            "arriveTime": datetime.combine(row.departure_date, datetime.min.time(), tzinfo=timezone.utc),
            "durationMinutes": 0, "stops": 0, "cabin": row.cabin, "baggage":"Check airline fare rules",
            "fare":{"base":float(row.base_fare),"taxes":float(row.taxes),"fees":float(row.fees),"total":float(row.total_fare)},
            "priceRating":"average","source":row.source,"lastUpdated":row.observed_at
        }
    # Stable demo detail for IDs produced by the fallback.
    parts = flight_id.split("-")
    if len(parts) >= 4:
        origin, dest, day = parts[1][:3], parts[1][3:], "-".join(parts[3:])
        flights = demo_flights(origin, dest, day, 8)
        return next((x for x in flights if x["id"] == flight_id), flights[0])
    raise HTTPException(404, "Flight not found")

@router.get("/fares")
def fare_history(route: str = Query(..., pattern=r"^[A-Z]{3}-[A-Z]{3}$"), db: Session = Depends(get_db)):
    rows = db.execute(select(FareObservation).where(FareObservation.route == route).order_by(FareObservation.observed_at.desc()).limit(500)).scalars().all()
    values = [float(x.total_fare) for x in rows]
    if not values:
        history = demo_history(route)
        values = [x["price"] for x in history]
    else:
        # Aggregate DB observations by observation date for a clean chart.
        grouped = {}
        for r in rows:
            grouped.setdefault(r.observed_at.date().isoformat(), []).append(float(r.total_fare))
        history = [{"date":d,"price":round(sum(v)/len(v))} for d,v in sorted(grouped.items())]
    current = values[0]
    avg = round(sum(values)/len(values))
    low, high = round(min(values)), round(max(values))
    sorted_values = sorted(values)
    percentile = round(sorted_values.index(current)/max(1,len(sorted_values)-1)*100)
    return {"route":route,"count":len(values),"history":history,"stats":{"current":round(current),"average":avg,"lowest":low,"highest":high,"percentile":percentile}}

@router.get("/price-history")
def price_history(route: str = Query(..., pattern=r"^[A-Z]{3}-[A-Z]{3}$"), db: Session = Depends(get_db)):
    return fare_history(route, db)

File: frontend/aeroindex/src/mock/flights.js
import { AIRLINES } from "./reference.js";
import { seededRandom, randRange } from "../lib/prng.js";

function rateFare(total, avg) {
  const ratio = total / avg;
  if (ratio < 0.85) return "very_good";
  if (ratio < 0.97) return "good";
  if (ratio < 1.1) return "average";
  if (ratio < 1.25) return "high";
  return "very_high";
}

function pad(n) {
  return String(n).padStart(2, "0");
}

/**
 * Generates a deterministic-ish list of mock flights for a route + date.
 * @param {{from:string, to:string, date:string}} params
 * @returns {import('../types').Flight[]}
 */
export function generateFlights({ from, to, date }) {
  const rng = seededRandom(`${from}-${to}-${date}`);
  const routeAvg = randRange(rng, 4200, 7800);
  const count = 5 + Math.floor(rng() * 4);
  const flights = [];

  for (let i = 0; i < count; i++) {
    const airline = AIRLINES[Math.floor(rng() * AIRLINES.length)];
    const depHour = 5 + Math.floor(rng() * 17);
    const depMin = Math.floor(rng() * 12) * 5;
    const durationMinutes = 90 + Math.floor(rng() * 90);
    const stops = rng() > 0.78 ? 1 : 0;
    const base = routeAvg * randRange(rng, 0.72, 1.35);
    const taxes = base * 0.14;
    const fees = 199 + Math.floor(rng() * 150);
    const total = Math.round(base + taxes + fees);

    const departTime = `${date}T${pad(depHour)}:${pad(depMin)}:00`;
    const arriveDate = new Date(new Date(departTime).getTime() + (durationMinutes + stops * 45) * 60000);

    flights.push({
      id: `${airline.code}-${from}${to}-${i}-${date}`,
      airlineCode: airline.code,
      flightNumber: `${airline.code} ${100 + Math.floor(rng() * 899)}`,
      from,
      to,
      departTime,
      arriveTime: arriveDate.toISOString(),
      durationMinutes: durationMinutes + stops * 45,
      stops,
      cabin: "Economy",
      baggage: "15kg check-in + 7kg cabin",
      fare: {
        base: Math.round(base),
        taxes: Math.round(taxes),
        fees,
        total,
      },
      priceRating: rateFare(total, routeAvg),
      source: rng() > 0.5 ? "Airline Direct" : "OTA Aggregate",
      lastUpdated: new Date(Date.now() - Math.floor(rng() * 45) * 60000).toISOString(),
    });
  }

  return flights.sort((a, b) => new Date(a.departTime) - new Date(b.departTime));
}

/**
 * @returns {import('../types').PriceHistoryPoint[]}
 */
export function generatePriceHistory(routeKey, days = 60) {
  const rng = seededRandom(routeKey);
  const avg = randRange(rng, 4200, 7800);
  const points = [];
  let cursor = avg * randRange(rng, 0.9, 1.1);

  for (let i = days; i >= 0; i--) {
    cursor += (rng() - 0.5) * avg * 0.06;
    cursor = Math.max(avg * 0.6, Math.min(avg * 1.6, cursor));
    const d = new Date();
    d.setDate(d.getDate() - i);
    points.push({ date: d.toISOString().slice(0, 10), price: Math.round(cursor) });
  }
  return points;
}

export function priceStats(history) {
  const values = history.map((p) => p.price);
  const current = values[values.length - 1];
  const avg = Math.round(values.reduce((a, b) => a + b, 0) / values.length);
  const low = Math.min(...values);
  const high = Math.max(...values);
  const sorted = [...values].sort((a, b) => a - b);
  const percentile = Math.round((sorted.indexOf(current) / (sorted.length - 1)) * 100);
  return { current, average: avg, lowest: low, highest: high, percentile };
}
